test/common.py

```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
+            """
+            # 'constraint' = positive constraints (must appear), 'avoid' = negative constraints (must not appear)
+            for constraint_type in ["constraints", "avoid"]:
+                constrained_sources = []
+                with open(test_source_path) as source_inp, open(out_path) as system_out:
+                    for sentno, (source, target) in enumerate(zip(source_inp, system_out)):
+                        target_words = target.rstrip().split()
+                        target_len = len(target_words)
+                        new_source = {'text': source.rstrip()}
+                        # From the odd-numbered sentences that are not too long, create constraints. We do
+                        # only odds to ensure we get batches with mixed constraints / lack of constraints.
+                        if target_len > 0 and sentno % 2 == 0:
+                            start_pos = 0
+                            end_pos = min(target_len, 3)
+                            constraint = ' '.join(target_words[start_pos:end_pos])
+                            new_source[constraint_type] = [constraint]
+                        constrained_sources.append(json.dumps(new_source))
+
+                new_test_source_path = os.path.join(work_dir, "test_constrained.txt")
+                with open(new_test_source_path, 'w') as out:
+                    for json_line in constrained_sources:
+                        print(json_line, file=out)
+
+                out_path_constrained = os.path.join(work_dir, "out_constrained.txt")
+                params = "{} {} {} --json-input".format(sockeye.translate.__file__,
+                                                        _TRANSLATE_PARAMS_COMMON.format(model=model_path,
+                                                                                        input=new_test_source_path,
+                                                                                        output=out_path_constrained,
+                                                                                        quiet=quiet_arg),
+                                                        translate_params)
+
+                with patch.object(sys, "argv", params.split()):
+                    sockeye.translate.main()
+
+                print('JSON_INPUT', open(new_test_source_path).readlines(),
+                      'CONSTRAINED', open(out_path_constrained).readlines(),
+                      'UNCONSTRAINED', open(out_path).readlines(), sep='\n')
+                for json_input, constrained_out, unconstrained_out in zip(open(new_test_source_path), open(out_path_constrained), open(out_path)):
+                    jobj = json.loads(json_input)
+                    print('JSON_INPUT', json_input, 'CONSTRAINED', constrained_out, 'UNCONSTRAINED', unconstrained_out, sep='\n')
+                    if jobj.get(constraint_type, None) == None:
+                        assert constrained_out == unconstrained_out
```

Comment:
would it be helpful to add error messages to these `assert` statements?

Reply:
I can add comments, but none of the other asserts have comments, and I'm not sure exactly what to put...

Reply:
lets leave as is, was just wondering about whether the if cases are clear, and/or what we are asserting. but after a second look its clear to me.